tfx/benchmarks/benchmark_base.py

```diff
@@ -19,22 +19,82 @@
 
 from absl import flags
 import apache_beam as beam
+from apache_beam.pipeline import PipelineOptions
 from tensorflow.python.platform import test  # pylint: disable=g-direct-tensorflow-import
+from tfx.benchmarks import mode_config
 
 FLAGS = flags.FLAGS
 flags.DEFINE_string(
     "beam_runner", "DirectRunner",
     "Beam runner to use - any runner name accepted by "
     "apache_beam.runners.create_runner")
 
-
 class BenchmarkBase(test.Benchmark):
+  """Base class for running Beam pipelines on various runners"""
 
-  def _create_beam_pipeline(self):
+  def __init__(self):
+    super(BenchmarkBase, self).__init__()
+    self.beam_pipeline_mode = mode_config.DEFAULT_MODE
+    self.num_workers = 1
+    self.cloud_dataflow_temp_loc = None
+
+  def _set_cloud_dataflow_flags(self):
+    self.flags = ["--runner=DataflowRunner",
+                  "--project=tfx-keshav",
+                  "--temp_location=" + self.cloud_dataflow_temp_loc,
+                  "--num_workers=%d" % self.num_workers,
+                  "--max_num_workers=" + self.num_workers,
+                  "--no_pipeline_type_check",
+                  "--setup_file=./setup.py",
+                  "--autoscaling_algorithm=NONE",
+                  "--region=us-central1"]
+
+  def _set_flink_on_k8s_operator_flags(self):
+    self.flags = ["--runner=PortableRunner",
+                  "--job_endpoint=localhost:8099",
+                  "--artifact_endpoint=localhost:8098",
+                  "--environment_type=EXTERNAL",
+                  "--environment_config=localhost:50000",
+                  "--parallelism=%d" % self.num_workers,
+                  "--no_pipeline_type_check"]
+
+  def _set_local_scaled_execution_flags(self):
+    self.flags = ["--runner=DirectRunner",
+                  "--direct_num_workers%d" % self.num_workers,
+                  "--direct_running_mode=multi_processing",
+                  "--no_pipeline_type_check"]
+
+  def set_beam_pipeline_mode(self, beam_pipeline_mode):
+    assert beam_pipeline_mode in mode_config.modes
+    self.beam_pipeline_mode = beam_pipeline_mode
+
+  def set_num_workers(self, num_workers):
+    self.num_workers = num_workers
+
+  def set_cloud_dataflow_temp_loc(self, cloud_dataflow_temp_loc):
+    self.cloud_dataflow_temp_loc = cloud_dataflow_temp_loc
+
+  def _create_beam_pipeline_default(self):
     # FLAGS may not be parsed if the benchmark is instantiated directly by a
     # test framework (e.g. PerfZero creates the class and calls the methods
     # directly)
     runner_flag = (
         FLAGS.beam_runner
         if FLAGS.is_parsed() else FLAGS["beam_runner"].default)
     return beam.Pipeline(runner=beam.runners.create_runner(runner_flag))
+
+  def _create_beam_pipeline(self):
```

Comment:
I'm concerned about the changes to the class API:

- This entangles details about Cloud Dataflow etc. into the Benchmark classes. This is not great because the benchmarks are about defining the _pipelines_ themselves rather than the _execution environment_. Keeping those two separate is one of the main philosophies of Beam.
- The runner configuration API is cumbersome: it allows users to set parameters that are unrelated to the runner e.g. I can call `set_cloud_dataflow_temp_loc` and then run the pipeline on Flink.

One way to maintain this separation is to add a `pipeline_options` parameter to `create_beam_pipeline` with a default value of `None`. Within `create_beam_pipeline`, pass those options into `beam.Pipeline`. Then provide a pure functions for each runner elsewhere that construct the appropriate `PipelineOptions`, with parameters appropriate for that runner.

What do you think? This would require quite a bit of restructuring so I'm OK with leaving it out of this PR.